website/docs/zkvm/developer-guide/optimization.md

```diff
@@ -0,0 +1,436 @@
+
+RISC Zero’s zkVM is designed and built to act like a physical CPU.
+We did this so you could use general purpose languages (e.g. Rust) and general purpose tools (e.g. Cargo, LLVM). **Generally, you can apply general purpose optimization techniques too!**
+
+In this guide, we’ll cover optimization of [zkVM guest programs].
+We’ll discuss when you can apply common techniques from other contexts, and when you need a new approach.
+
+[zkVM guest programs]: ./guest-code-101
+
+## Background
+
+### What is the zkVM, really?
+
+**The zkVM is essentially a CPU.**
+
+In particular, it is an implementation of the [RISC-V architecture] (specifically riscv32im), similar to how your laptop’s CPU may be an implementation of an [x86] or [ARM] architecture.
+This biggest difference is that the zkVM is implemented with [arithmetic circuits], in software, instead of circuitry made from silicon and copper.
+
+[RISC-V architecture]: /reference-docs/about-risc-v
+[x86]: https://en.wikipedia.org/wiki/X86
+[ARM]: https://en.wikipedia.org/wiki/ARM_architecture_family
+[arithmetic circuits]: https://dev.risczero.com/proof-system/stark-by-hand
+
+### What is a “cycle”?
+
+Both in the zkVM and on physical CPUs, the cost of an operation is measured in “clock cycles”.
+
+Intuitively, a "clock cycle" is the smallest unit of time in a CPU's operation, representing one tick of the CPU's internal clock and the time it takes to execute a basic CPU operation (e.g. adding two integers).
+We’ll refer to this as a “cycle”. [^1]
+
+**Proving times for the zkVM are directly related to the number of cycles in an execution.**
+
+## General techniques and advice
+
+**Start by applying general techniques and best practices for optimizing your code.**
+
+[The Rust Performance Book] is great resource.
+It’s not very long, it covers a range of topics important to performance, and gives applicable advice for optimization.
+If you are new to optimization, or new to Rust, we recommend you read this guide.
+
+[The Rust Performance Book]: https://nnethercote.github.io/perf-book/
+
+### Don’t assume, measure.
+
+Performance is complex, in the zkVM as on a physical CPU.
+Don’t assume you know what the bottlenecks are. Measure and experiment.
+
+If you make a function 100x faster (or ∞x faster) but it only takes 1% of your execution time, you’ll see a &lt;1% improvement in performance.
+This is generally referred to as [Amdahl’s Law](https://en.wikipedia.org/wiki/Amdahl%27s_law), and practically it means you shouldn't waste your time optimizing something that's not the taking a significant portion of execution time.
+
+### Measuring by printing to console
+
+Starting simple, measure by adding an `eprintln!` line to your guest code to measure how long an operation takes, and how many times it is called.
+
+Using [`env::get_cycle_count()`] will tell you the current number of execution cycles that have occurred in your program.
+
+As an example:
+
+```rust title="methods/guest/src/main.rs"
```

Comment:
In the past we were including comments with versioning tags like `v0.19` to make it easier to handle version maintenance via `ctrl-f`. @flaub may have other thoughts on how to handle tagging code to help manage maintenance